### PR TITLE
DOC: fix typos found by codespell

### PR DESCRIPTION
**Pull request by DimitriPapadopoulos**
#### What does this implement/fix?

Fix typos found by [codespell](https://github.com/codespell-project/codespell).

#### Additional information

Hopefully, I have not modified vendored files that should be fixed upstream.

### CODE REVIEW

**Comment by rkern** [comment]
Please see our comments in #19041 about how to structure this kind of PR for review in manageable chunks. I'm afraid that doing it all at once is not tenable for us.

File: scipy/linalg/tests/test_lapack.py [hunk 1]
```diff
@@ -662,8 +662,8 @@ def test_lartg():
 
         if np.iscomplexobj(g):
             assert_allclose(sn, -4.0j/5.0)
-            assert_(type(r) == complex)
-            assert_(type(cs) == float)
+            assert_(isinstance(r, complex))
+            assert_(isinstance(cs, float))
```
Comment (lucascolley) [inline scipy/linalg/tests/test_lapack.py:666]: this change would probably be better in a separate PR - it does look like a Flake8 improvement, but not a typo.

File: scipy/optimize/_shgo_lib/_complex.py [hunk 2]
```diff
@@ -137,8 +137,7 @@ def __init__(self, dim, domain=None, sfield=None, sfield_args=(),
             self.min_cons = constraints
             self.g_cons = []
             self.g_args = []
-            if (type(constraints) is not tuple) and (type(constraints)
-                                                     is not list):
+            if not isinstance(constraints, tuple) and not isinstance(constraints, list):
```
Comment (lucascolley) [inline scipy/optimize/_shgo_lib/_complex.py:140]: same comment about 'not a typo' here.

File: scipy/optimize/tests/test__shgo.py [hunk 3]
```diff
@@ -27,7 +27,7 @@ def __init__(self, bounds, expected_x, expected_fun=None,
 def wrap_constraints(g):
     cons = []
     if g is not None:
-        if (type(g) is not tuple) and (type(g) is not list):
+        if not isinstance(g, tuple) and not isinstance(g, list):
```
Comment (lucascolley) [inline scipy/optimize/tests/test__shgo.py:30]: and here (see above comments)

File: scipy/special/_precompute/wright_bessel.py [hunk 4]
```diff
@@ -265,7 +265,7 @@ def fp(eps, a, b, x, phi):
     def arclength(eps, a, b, x, epsrel=1e-2, limit=100):
         """Compute Arc length of f.
 
-        Note that the arg length of a function f fro t0 to t1 is given by
+        Note that the arg length of a function f from t0 to t1 is given by
```
Comment (lucascolley) [inline scipy/special/_precompute/wright_bessel.py:268]: There is another typo here: `arg` -> `arc`

File: scipy/stats/_page_trend_test.py [hunk 5]
```diff
@@ -348,7 +348,7 @@ def page_trend_test(data, ranked=False, predicted_ranks=None, method='auto'):
                              f"from 1 to {n} (the number of columns in "
                              f"`data`) exactly once.")
 
-    if type(ranked) is not bool:
+    if not isinstance(ranked, bool):
```
Comment (lucascolley) [inline scipy/stats/_page_trend_test.py:351]: and here (see above comments)

File: scipy/stats/tests/test_mstats_basic.py [hunk 6]
```diff
@@ -1123,7 +1123,7 @@ def test_kurtosistest_result_attributes(self):
         check_named_results(res, attributes, ma=True)
 
     def test_regression_9033(self):
-        # x cleary non-normal but power of negtative denom needs
+        # x clearly non-normal but power of negtative denom needs
```
Comment (lucascolley) [inline scipy/stats/tests/test_mstats_basic.py:1126]: another typo: `negtative` -> `negative`.

File: scipy/stats/tests/test_stats.py [hunk 7]
```diff
@@ -5760,7 +5760,7 @@ def test_normalitytests():
     assert_raises(ValueError, stats.normaltest, x, nan_policy='raise')
     assert_raises(ValueError, stats.normaltest, x, nan_policy='foobar')
 
-    # regression test for issue gh-9033: x cleary non-normal but power of
+    # regression test for issue gh-9033: x clearly non-normal but power of
     # negtative denom needs to be handled correctly to reject normality
```
Comment (lucascolley) [inline scipy/stats/tests/test_stats.py:5764]: `negtative` -> `negative` here too.

File: tools/refguide_check.py [hunk 8]
```diff
@@ -205,7 +205,7 @@ def find_names(module, names_dict):
             res = re.match(pattern, line)
             if res is not None:
                 name = res.group(1)
-                entry = '.'.join([module_name, name])
+                ## entry = '.'.join([module_name, name])
```
Comment (lucascolley) [inline tools/refguide_check.py:208]: it looks like this line can just be removed?

**Review by lucascolley — COMMENTED** [review]
Hi, see my comments about maybe saving the `isinstance` changes for a separate PR, and the `entry` line in `refguide_check`.

File: scipy/linalg/tests/test_lapack.py (hunk 1 above)
Comment (DimitriPapadopoulos) [reply scipy/linalg/tests/test_lapack.py:666]: This is a separate commit, attempting to fix linter errors in CI jobs.

File: scipy/optimize/_shgo_lib/_complex.py (hunk 2 above)
Comment (DimitriPapadopoulos) [reply scipy/optimize/_shgo_lib/_complex.py:140]: Same here.

File: scipy/optimize/tests/test__shgo.py (hunk 3 above)
Comment (DimitriPapadopoulos) [reply scipy/optimize/tests/test__shgo.py:30]: And same here.

File: scipy/stats/_page_trend_test.py (hunk 5 above)
Comment (DimitriPapadopoulos) [reply scipy/stats/_page_trend_test.py:351]: Same.

File: tools/refguide_check.py (hunk 8 above)
Comment (DimitriPapadopoulos) [reply tools/refguide_check.py:208]: Indeed, I commented it to please the linter.

File: scipy/special/_precompute/wright_bessel.py (hunk 4 above)
Comment (DimitriPapadopoulos) [reply scipy/special/_precompute/wright_bessel.py:268]: This one is difficult to catch without processing n-grams instead of isolated words.

**Comment by DimitriPapadopoulos** [comment]
@lucascolley I have added the `isinstance` changes to please the CI linter. Not sure how to avoid the CI complaining without the `isinstance` commit.

**Comment by lucascolley** [comment]
> I have added the isinstance changes to please the CI linter. Not sure how to avoid the CI complaining without the isinstance commit.

In that case, I would recommend submitting a PR for these changes before submitting a PR for typos in files where it is complaining (and split the PRs up as Robert has requested).

**Comment by rkern** [comment]
At the moment, this LGTM, more or less. I'd make a different judgment call to leave `re-use` and `co-ordinates` alone, but I'm fine with removing the hyphens.

Please don't force-push when working with hard-to-review PRs like this one. There's no way for us to know which changes are new from the last time we looked it, so we have to examine each line again. If you really do need to make another change, please just add more commits to the end like normal.

**Comment by DimitriPapadopoulos** [comment]
I had to rebase to include the linter fixes from https://github.com/scipy/scipy/pull/19296.

**Comment by DimitriPapadopoulos** [comment]
Because "reuse" and "coordinate" are words in the dictionary, they don't need the hyphen. You could write "re-use" to stress the "re", but I don't see the benefit of "co-ordinate".

All the spellchecking tools I use flag the hyphen as a misspelling. I don't mean to say that  "re-use" and "coordinate" are not English words - after all the latter can be [found in the OED](https://www.oed.com/search/dictionary/?q=co-ordinate).  It's just that the words without hyphen are really more common and standard:
https://books.google.com/ngrams/graph?content=co-ordinate%2Ccoordinate&year_start=1800&year_end=2019&corpus=en-2019

File: scipy/optimize/_trustregion_constr/tests/test_qp_subproblem.py [hunk 9]
```diff
@@ -558,7 +558,7 @@ def test_inactive_box_constraints(self):
         assert_array_almost_equal(x, x_kkt)
 
     # The box constraints active and the termination is
-    # by maximum iterations (infeasible iteraction).
+    # by maximum iterations (infeasible iteration).
```
Comment (h-vetinari) [inline scipy/optimize/_trustregion_constr/tests/test_qp_subproblem.py:561]: This and below should almost certainly be "interaction"

File: scipy/stats/_unuran/unuran_callback.h [hunk 10]
```diff
@@ -4,7 +4,7 @@
 
 #define UNURAN_THUNK(CAST_FUNC, FUNCNAME, LEN)                                              \
     PyGILState_STATE gstate = PyGILState_Ensure();                                          \
-    /* If an error has occured, return INFINITY. */                                         \
+    /* If an error has occurred, return INFINITY. */                                         \
```
Comment (h-vetinari) [inline scipy/stats/_unuran/unuran_callback.h:7]: Please re-align line-continuation

File: scipy/_lib/_uarray/__init__.py [hunk 11]
```diff
@@ -33,7 +33,7 @@
 creating multimethods are mostly orthogonal activities, and knowing
 one doesn't necessarily require knowledge of the other, although it
 is certainly helpful. We expect core API designers/specifiers to write the
-multimethods, and implementors to override them. But, as is often the case,
+multimethods, and implementers to override them. But, as is often the case,
```
Comment (h-vetinari) [inline scipy/_lib/_uarray/__init__.py:36]: Still not enthusiastic about this. I'd prefer to not warn on rare words by default, but then again, if we're only talking about one legitimate occurrence of `-ors` (the convertor one is even more of an outlier), then it's not all that relevant. 🤷

**Review by h-vetinari — COMMENTED** [review]
Generally in favour of this. Some of the typos are pretty egregious. 😅 

If the price of automation to avoid that in the future is removing the hyphen from "re-use" and changing "implementor" to "implementer", then I think that's a worthwhile trade-off.

Pretty sure I found one instance of a wrong correction though.

File: scipy/optimize/_trustregion_constr/tests/test_qp_subproblem.py (hunk 9 above)
Comment (DimitriPapadopoulos) [reply scipy/optimize/_trustregion_constr/tests/test_qp_subproblem.py:561]: My wrong. I had to choose between these two words, and I have to admit I don't understand the comment – I probably lack context.

File: scipy/_lib/_uarray/__init__.py (hunk 11 above)
Comment (DimitriPapadopoulos) [reply scipy/_lib/_uarray/__init__.py:36]: That's OK, I can revert this one. In the event of automation of spell checking later on, such words can be added to an allowlist.

That said, I believe it's beneficial to use spelling that is as standard as possible when writing documentation, both for readers and to avoid alerts from spell checkers such as _aspell_. However, that's just my opinion.

**Review by rgommers — APPROVED** [review]
This looks like it's in good shape; all comments were addressed. Let's get this in. Thanks @DimitriPapadopoulos & all reviewers.
